### ISSUE TITLE
Dropout layer misplaced in output?

### ISSUE THREAD

**Issue by rastna12** [ronaldseoh/ronald_bdl#1]
Hello,
        I have been reading your paper 'Qualitative Analysis of Monte Carlo Dropout' recently and reading through your code. Many thanks for making it available. I noticed something odd in line 88 of fcnet.py. See below link. Adding a dropout layer after the last linear layer seems odd to me. I have typically seen it placed _before_ the final linear output layer, not after. See Yarin Gal's MC dropout model definition [here](https://github.com/yaringal/DropoutUncertaintyExps/blob/master/net/net.py). The dropout precedes the final linear layer. In fcnet.py's case, the model output is 1-D, which is then being dropped periodically. This appears to be incorrectly inflating the computed variance values, especially for models that produce large outputs. 

If this is the case, then this issue can be fixed by swapping these two lines. Thoughts?

https://github.com/ronaldseoh/ronald_bdl/blob/9485d2ac4ffa18684d279bd2cdffd24888719f5f/ronald_bdl/models/fcnet.py#L88

**Comment by ronaldseoh** [comment]
Hi! Thank you so much for taking the time to read my paper and codes! This was my class project last year, so please be aware that there might be some serious errors as this didn't go through any rigorous review process.

That said, I think I do remember noticing that from Gal's code but still choosing to put dropout after the final linear layer because I simply wasn't able to find the exact reason why I should not - I mean, aren't the final layer's weights also model parameters?  I wasn't sure why I should assume different distribution for them.

I moved on to other topics for now, so what I said above is quite likely to be a complete nonsense. I will try to come back to this after checking Gal's paper again over the weekend. If you do know the answer, I'd really appreciate it if you could share it with me here!

**Comment by rastna12** [comment]
Thanks for getting back so quickly. My primary source for this is just
Gal's paper 'Dropout as a Bayesian approximation: Representing model
uncertainty in deep learning' in Section 3 where he says "...with dropout
applied *before *every weight layer...". I am not an expert in this area,
but my thinking at the moment is because dropout is effectively breaking
the connection to the next node in the sequence. If the dropout occurs
after the final linear layer calculation is performed, then the model will
periodically return zero as a final answer, which just doesn't feel right.
That layer's output is not being passed into another layer, so we don't
need to randomly drop it. It's just being passed into our loss function or
stored as the trained model's output.

I've been finding and collecting examples of MC dropout code and running
them to compare their output to a model I'm working on myself. If I make
that tweak to your code then the results fall in line with the others.

-rps

**Comment by ronaldseoh** [comment]
So I had a look at the paper again and you are right - Under Gal's proof (check [the appendix](https://arxiv.org/abs/1506.02157) of the paper), it is the units of each weight matrices and corresponding inputs being dropped, so having additional dropout layer is not part of the equivalence. I usually thought of dropouts as each layer's *outputs* being dropped, so I guess the things got mixed up in my head at some point :cry:

Plus, I've also realized that we shouldn't have the bias term in the final output layer to follow his formulation exactly. [PyTorch `Linear` layer](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html#torch.nn.Linear) and [Keras `Dense` layer](https://keras.io/api/layers/core_layers/dense/) apparently adds bias term by default.

**Comment by ronaldseoh** [comment]
I will have to re-run my experiments and revise the paper pretty soon, but I think this probably (hopefully) won't change most of the observations I've made there. Variance values will get smaller though.

**Comment by rastna12** [comment]
That's an interesting and good observation about the bias. Do you follow
why Gal does not use a bias term in his paper? The only detail I found is a
footnote in the appendix that says: "Note that we omit the outer-most bias
term as this is equivalent to centring the output". However, looking at
Gal's code he does use a bias term. I guess leaving out the bias term
forces the last layer to only be able to linearly shear and stretch the
output since shifting and translation is handled by the bias. I'm not sure
what the significance of that tweak is in relation to the broader theory.

Regarding your model, couldn't another dropout layer be added before the
'noise' linear layer if the model is predicting the heteroscedastic noise?
That would match the pattern with the 'activation' layer at the very least.
I suppose that would give us a better sample of the posterior distribution
if I follow Gal's math correctly?

Ryan

**Comment by ronaldseoh** [comment]
Hello Ryan, apologies for the late reply. I'm currently working on a few different things and haven't really had chance to take a proper look at this lately :( If our discussion is no longer relevant to your project, please ignore this message.

I think the easiest explanation for the last bias term is that they never get 'dropped out' in any way under our formulation. Unlike other trainable weights where they contributions to the output would get dropped out at some point, this last bias goes straight to the output. As a result, it would shift the predictive distribution, but doesn't actually contribute to the overall shape of the distribution.

Regarding your second point, you might want to check out [Gal's Github repo about heteroscedastic noise](https://github.com/yaringal/HeteroscedasticDropoutUncertainty). The idea here is that we assume our noise to be some linear function of each data point. Since this noise is not entirely caused by our main model, we do not try to estimate the distribution of these weights. By the way, this is where I found thinking about epistemic and aleatoric certainty to be really interesting and confusing at the same time: If we get to observe some data, how much of it should be considered as  'noise' and actual 'truth'?
